src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -221,56 +228,79 @@ object DeployPlugin extends AutoPlugin {
+    val envConfFile = new File(universalStagingDir, s"conf/$deployEnv.conf")
     if (envConfFile.exists) {
-      log.info(s"Copying config for ${deployEnv} . . .")
+      log.info(s"Copying config for $deployEnv . . .")
       val destConfFile = new File(universalStagingDir, "conf/env.conf")
       IO.copyFile(envConfFile, destConfFile)
     } else {
       log.info("")
-      log.info(s"WARNING: Could not find config file ${deployEnv}.conf!")
+      log.info(s"WARNING: Could not find config file '$deployEnv.conf'!")
       log.info("")
       log.info("Press ENTER to continue with no environment configuration, CTRL-C to abort.")
       log.info("")
       System.console.readLine()
     }
 
-    // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh
-    // operations.
-    val sshCommand = {
-      val pemEnvVar = "AWS_PEM_FILE"
-      val sshKeyfile = sys.env(pemEnvVar)
-      val sshUser = configMap("deploy.user.ssh_username")
-      Seq("ssh", "-i", sshKeyfile, "-l", sshUser)
+    val deploys = deployConfig.hostConfigs map { hostConfig =>
+      Future {
+        // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh
+        // operations.
+        val sshCommand = {
+          val sshKeyfile = sys.env("AWS_PEM_FILE")
+          val sshUser = hostConfig.sshUser
+          Seq("ssh", "-i", sshKeyfile, "-l", sshUser)
+        }
+        val deployHost = hostConfig.host
+        val deployDirectory = hostConfig.directory
+
+        val rsyncDirs = deployDirs.value map (name => s"--include=/$name")
+        val rsyncCommand = Seq.concat(
+          Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")),
+          rsyncDirs,
+          Seq(
+            "--exclude=/*",
+            "--delete",
+            universalStagingDir.getPath + "/",
+            deployHost + ":" + deployDirectory
+          )
+        )
+
+        // Shell-friendly version of rsync command, with rsh value quoted.
+        val quotedRsync = rsyncCommand.patch(
+          2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
+        ).mkString(" ")
+        log.info("Running " + quotedRsync + " . . .")
+        if (Process(rsyncCommand).! != 0) {
+          throw new IllegalArgumentException(s"Error running rsync to host '$deployHost'.")
+        }
+
+        // Now, ssh to the remote host and run the restart script.
+        val restartScript = deployDirectory + "/" + hostConfig.startupScript
+        val restartCommand = Seq.concat(
+          sshCommand,
+          Seq(deployHost, restartScript, "restart"),
+          hostConfig.startupArgs map { "--" +: _ } getOrElse Seq()
+        )
+        log.info("Running " + restartCommand.mkString(" ") + " . . .")
+        if (Process(restartCommand).! != 0) {
+          throw new IllegalArgumentException(s"Error running restart command on host '$deployHost'.")
+        }
+      } map Success.apply recover { case e => Failure(e) } // Map to Try so we can fold over errors.
```

Comment:
It would be clearer to just wrap the body of the `Future.apply` call in `Try.apply`.